src/transformers/modeling_longformer.py

```diff
@@ -0,0 +1,667 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch Longformer model. """
+
+
+import logging
+import math
+from functools import lru_cache
+import torch
+import torch.nn as nn
+from torch.nn import functional as F
+from torch.nn import CrossEntropyLoss
+
+from .configuration_longformer import LongformerConfig
+from .file_utils import add_start_docstrings, add_start_docstrings_to_callable
+from .modeling_roberta import RobertaModel, RobertaLMHead
+from .modeling_bert import BertPreTrainedModel
+
+
+logger = logging.getLogger(__name__)
+
+# TODO: upload model and config files
+LONGFORMER_PRETRAINED_MODEL_ARCHIVE_MAP = {
+    "longformer-base-4096": "https://ai2-s2-research.s3-us-west-2.amazonaws.com/longformer/longformer-base-4096/pytorch_model.bin",
+    "longformer-large-4096": "https://ai2-s2-research.s3-us-west-2.amazonaws.com/longformer/longformer-large-4096/pytorch_model.bin",
+}
+
+
+class LongformerSelfAttention(nn.Module):
+    def __init__(self, config, layer_id):
+        super(LongformerSelfAttention, self).__init__()
+        if config.hidden_size % config.num_attention_heads != 0:
+            raise ValueError(
+                "The hidden size (%d) is not a multiple of the number of attention "
+                "heads (%d)" % (config.hidden_size, config.num_attention_heads))
+        self.output_attentions = config.output_attentions
+        self.num_heads = config.num_attention_heads
+        self.head_dim = int(config.hidden_size / config.num_attention_heads)
+        self.embed_dim = config.hidden_size
+
+        self.query = nn.Linear(config.hidden_size, self.embed_dim)
+        self.key = nn.Linear(config.hidden_size, self.embed_dim)
+        self.value = nn.Linear(config.hidden_size, self.embed_dim)
+
+        # separate projection layers for tokens with global attention
+        self.query_global = nn.Linear(config.hidden_size, self.embed_dim)
+        self.key_global = nn.Linear(config.hidden_size, self.embed_dim)
+        self.value_global = nn.Linear(config.hidden_size, self.embed_dim)
+
+        self.dropout = config.attention_probs_dropout_prob
+
+        self.layer_id = layer_id
+        attention_window = config.attention_window[self.layer_id]
+        assert attention_window % 2 == 0, f'`attention_window` for layer {self.layer_id} has to be an even value. Given {attention_window}'
+        assert attention_window > 0, f'`attention_window` for layer {self.layer_id} has to be positive. Given {attention_window}'
+
+        self.one_sided_attention_window_size = attention_window // 2
+
+    def _skew(self, x, direction, padding_value):
+        '''Convert diagonals into columns (or columns into diagonals depending on `direction`'''
+        x_padded = F.pad(x, direction, value=padding_value)
+        x_padded = x_padded.view(*x_padded.size()[:-2], x_padded.size(-1), x_padded.size(-2))
+        return x_padded
+
+    def _skew2(self, x, padding_value):
```

Comment:
does the `padding_value` correspond to the `config.pad_token_id` here?

Reply:
Not sure if that's the case. But if `padding_value` always corresponds to `config.pad_token_id` we could save it once in each class and access it from the class instead of using it as an argument to the functions.

Reply:
`padding_value` is a large negative value to set attention to zero after softmax. I will clarify this.